akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -1277,10 +1285,58 @@ class DDataShardCoordinator(
         stashGetShardHomeRequest(sender(), g) // must wait for update that is in progress
 
     case ShardCoordinator.Internal.Terminate =>
-      log.debug("Received termination message while waiting for update")
+      log.debug("The ShardCoordinator received termination message while waiting for update")
       terminating = true
       stash()
 
+    case RememberEntitiesCoordinatorStore.UpdateDone(shard) =>
+      require(shardId.contains(shard))
+      if (!waitingForStateWrite) {
+        log.debug("The ShardCoordinator saw remember shard start successfuly written {}", evt)
+        if (shardId.isDefined) timers.cancel(RememberEntitiesTimeoutKey)
+        unbecomeAfterUpdate(evt, afterUpdateCallback)
+      } else {
+        log.debug("The ShardCoordinator saw remember shard start successfuly written {}, waiting for state update", evt)
+        context.become(
+          waitingForUpdate(
+            evt,
+            shardId,
+            waitingForStateWrite = true,
+            waitingForRememberShard = false,
+            afterUpdateCallback = afterUpdateCallback))
+      }
+
+    case RememberEntitiesCoordinatorStore.UpdateFailed(shard) =>
+      require(shardId.contains(shard))
+      log.error(
+        "The ShardCoordinator was unable to update shards distributed state, {} ({})",
+        if (terminating) "terminating" else "retrying",
+        evt)
+      if (terminating) context.stop(self)
+      else {
+        // retry
+        rememberShardAllocated(shardId.get)
+      }
+
+    case RememberEntitiesTimeout(shard) =>
```

Comment:
isn't this identical to `UpdateFailed` aside from the log message? do we need both?

Reply:
at least extract to a method that is used by both

Reply:
Maybe not that valuable as I was thinking with the different logs no